### ISSUE TITLE
Proposed projects

### ISSUE THREAD

**Issue by andreafioraldi** [AFLplusplus/LibAFL#119]
In this issue, I proposed several projects based on libafl (like libafl_frida) that we would be glad to include here.
As LibAFL is newly born there is a lot of work on the lib itself and we mostly work only on it, so we are seeking help for these projects.

- [ ] Extend Forkserver (#111) to work on windows including https://github.com/sslab-gatech/winnie/tree/master/forklib from Winnie in libafl_targets
- [ ] Frida ASan for other architectures, ATM it is just for arm64
- [ ] libafl_tinyinst to support TinyInst as instrumentation backend. This will require some TinyInst bindings to Rust before
- [ ] libafl_pt using libxdc as Intel PT decoder
- [ ] QEMU and unicorn support with snapshot fuzzing and sanitization. I already started it with https://github.com/AFLplusplus/qemu-libafl-bridge but feel free to ask for details about how I plan to integrate emulators in LibAFL if you want to jump in
- [ ] Anything with no_std like embedding a fuzzer in the kernel, we would love to see such usecase.

Then, if you want to implement any of the recent fuzzing techniques (https://wcventure.github.io/FuzzingPaper/ can be useful) feel free to ping us in order to know if we are already implementing the technique that you are interested in or not.

**Comment by bitwave** [comment]
see  #551 for full system libafl_qemu

**Comment by vanhauser-thc** [comment]
@syheliel it is actively maintained. it was integrated into afl++ just 1-2 months ago. check out the nyx stuff in afl++. otherwise sergej is very helpful if you have questions.

**Comment by syheliel** [comment]
@vanhauser-thc thanks!

**Comment by irshimul** [comment]
Highly interested in the project related to PANGOLIN mutator. I am curious to know what other recent academic research projects are being integrated into AFL++ at this moment. 

Can you please tell me what you use for communication (irc or slack)?

**Comment by tokatoka** [comment]
hello,
we are on this channel
https://github.com/AFLplusplus/AFLplusplus/issues/681

**Comment by Kanaintmeandyo** [comment]
hello, I am highly interested in contributing to the project. Please guide me through some basic starting steps.

**Comment by Srg213** [comment]
Hello.
I see that the issue is still open. I am interested in the project 'write user-level threads for libafl_qemu'. There is guide for beginners on discord but I could not find the status of this project. Can someone guide me through the materials and contributing steps?

**Comment by domenukk** [comment]
> Hello. I see that the issue is still open. I am interested in the project 'write user-level threads for libafl_qemu'. There is guide for beginners on discord but I could not find the status of this project. Can someone guide me through the materials and contributing steps?

Just a heads up, this is a rather complex project that requires deep understanding of OS internals and qemu. Of course I don't know your skill set, but if you don't have an idea where to get started, you may want to pick an easier goal such as #36 first.
For the libafl_qemu threading project, you would probably start by picking a simple threaded target and hooking it up in libafl_qemu, then go from there.

**Comment by bkrl** [comment]
Hello,

Is AFL++ planning to participate in GSoC this year? I'm interested in the project "Start rewriting AFL++'s afl-fuzz in Rust as a frontend of LibAFL."

**Comment by andreafioraldi** [comment]
> Hello,
> 
> Is AFL++ planning to participate in GSoC this year? I'm interested in the project "Start rewriting AFL++'s afl-fuzz in Rust as a frontend of LibAFL."

Yes we will partecipate

**Comment by bkrl** [comment]
Where would be best place to discuss potential GSoC projects? Should I just open a thread in the Discord?

**Comment by tokatoka** [comment]
Let's discuss here #1098

**Comment by satishsahu25** [comment]
is this issue open?

**Comment by juriSacchetta** [comment]
Hello everyone,

I'd like to inform the community that I began working on my thesis based on the project `Write user-level threads for libafl_qemu.` approximately one month ago. 
I will continue to provide updates and progress reports as I move forward.

**Comment by mlgiraud** [comment]
Hi, I'd just like to let you know that we are currently working on a unicorn based approach that will also make it possible to emulate multiple processes or threads in parallel. It is based on a custom kernel implementation that delegates all I/O to system components to which the fuzzer can individually supply inputs (i.e. the fuzzer acts as a stand-in for the actual component). This also makes multi-input-stream fuzzing possible (e.g. fuzzing a tcp stream in conjunction with udp inputs. This is relevant to for example RTSP).
The whole thing also implements copy on write, enabling the fuzzer to jump back to specific points to make stateful fuzzing easier.

Since this is a huge project however, it will take some more time as currently only a student of mine and I are working on it. Once we have a somewhat working version ready with a few of the basic pieces, we will most likely open source it and also be willing to integrate at least parts of it into libafl.

I would be interested in hearing if there is any overlap with already existing developments and parts in libAFL, since it is difficult keeping an overview over all the new stuff being published.

**Comment by addisoncrump** [comment]
@mlgiraud you've probably already seen #1617 and #913 -- these are probably closest to what you're doing. It may also be possible to avoid Unicorn/custom kernel entirely by developing a libafl kernel module (just use no_std and it should just work, though you will need an allocator) and intercepting these calls either by intercepting the system calls themselves, or by wrapping the creation of targeted sources of input by way of opening character devices defined by the kernel module. Hope this helps!

**Comment by mlgiraud** [comment]
@addisoncrump Yeah i saw your multipart input PR. I think it is not 100% suited for my use case but i will most likely adapt a few ideas from there. This will become clearer in the future when everything becomes more stable on our side. I'm not quite sure how #913 relates to my work though. Could you elaborate on why you think this might be relevant?

Regarding unicorn: We are using full CPU emulation, since we want to be able to emulate different architectures (e.g. ARM on x86). Of course an approach that skips emulation will be faster, but not as flexible, but that is not our goal here. The nice benefit we get here is also that we can decide which process to schedule, making it possible to e.g. further investigate concurrency bug detection via fuzzing (at least that's what im hoping for ;))

**Comment by addisoncrump** [comment]
#913 is relevant because it allows for the fuzz testing of programs which process sequences of inputs and buffer. For example, fuzzing a remote network target where each "input" (probably a higher-level packet of some sort) has a response. Doesn't make a ton of sense to send, wait, receive when you can repeatedly send and then asynchronously receive feedback. Since you're dead-set on using Unicorn, this probably doesn't make sense for your use case as I originally believed.

**Comment by Skryptonyte** [comment]
Hi, is there a place where I can communicate with the mentors about a project idea. Previous years had separate issues made for each GSoC iteration but I'm unable to find one for 2024.

**Comment by kd1729** [comment]
Hi, I would like to take up the ideas for GSOC, what is the procedure for the same? Any contributing guidelines? How to contact the mentor and get your proposal reviewed?

**Comment by domenukk** [comment]
Ideally candidates will work on github issues before the proposal deadline to show us their engineering skills, and talk to us about which projects they are interested in. Then we decide on candidates according to how confident we are they will be able to finish projects successfully. Happy hacking :)

**Comment by maxammann** [comment]
Any new ideas for 2024?

**Comment by domenukk** [comment]
https://github.com/AFLplusplus/LibAFL/issues/2964

**Comment by domenukk** [comment]
Superseeded by #2992
